lib/promise.js

```diff
@@ -1,17 +1,44 @@
 'use strict';
 
 var Promise = require('bluebird')
-  , _then = Promise.prototype._then;
-
-Promise.prototype._then = function (didFulfill, didReject, didProgress, receiver, internalData) {
-  if (Promise.Sequelize.cls) {
-    var ns = Promise.Sequelize.cls;
-    if (typeof didFulfill === 'function') didFulfill = ns.bind(didFulfill);
-    if (typeof didReject === 'function') didReject = ns.bind(didReject);
-    if (typeof didProgress === 'function') didProgress = ns.bind(didProgress);
-  }
-  
-  return _then.call(this, didFulfill, didReject, didProgress, receiver, internalData);
-};
-
-module.exports = Promise;
\ No newline at end of file
+  , shimmer = require('shimmer');
+
+// functionName: The Promise function that should be shimmed
+// fnArgs: The arguments index that should be CLS enabled (typically all callbacks). Offset from last if negative
+function shimCLS(object, functionName, fnArgs){
+  shimmer.wrap(object, functionName, function(fn) {
+    return function () {
+      if (Promise.Sequelize && Promise.Sequelize.cls) {
+        var ns = Promise.Sequelize.cls;
+        for(var x=0; x<fnArgs.length; x++) {
+          var argIndex = fnArgs[x] < 0 ? arguments.length + fnArgs[x] : fnArgs[x];
+          if ( argIndex < arguments.length && typeof arguments[argIndex] === 'function' ) {
+            arguments[argIndex] = ns.bind( arguments[argIndex] );
+          }
+        }
+      }
+
+      return fn.apply(this, arguments);
+    };
+  });
+}
+
+// Core
+shimCLS(Promise, 'join', [-1]);
+shimCLS(Promise.prototype, 'then', [0, 1, 2]);
+shimCLS(Promise.prototype, 'spread', [0, 1]);
```

Comment:
Not sure you need to shim `spread/catch/error/finally`.
Might want to add some tests so we can be sure though.